src/components/login/AuthenticatePage.js

```diff
@@ -0,0 +1,59 @@
+import React, { Component, Fragment } from 'react';
+import { connect } from 'react-redux';
+import { withRouter } from 'react-router-dom';
+import qs from 'qs';
+import { get } from 'lodash/fp';
+
+import Loader from 'components/common/Loader';
+
+import {
+  getUserTokenWithCode,
+  saveTokenToLocalStorage
+} from 'actions/user.actions';
+
+class AuthenticatePage extends Component {
+  componentDidMount() {
+    const searchParams = get('search', this.props.location);
+
+    if (searchParams) {
+      const codeParams = qs.parse(searchParams);
+      const userCode = codeParams['?code'];
+      this.props.getUserTokenWithCode(userCode);
+    }
+
+    if (!searchParams) {
+      this.props.token
+        ? this.props.history.push('/')
+        : this.props.history.push('/login');
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.props.badCodeRequest) {
```

Comment:
bad code request means the token was failed because of invalid code?
I think it will be better as a middleware flow

Reply:
The component sends the received code and gets a token back, if the token is invalid, it updates the "badCode" to true, AuthenticatePage listen to updates and will get either token to badCode and will redirect to a different page accordingly. what do you mean by a middleware flow?

Reply:
Consider this flow:

1. component mounts and reads code from url
2. component fires completeAuth api action with the code
3. api action onError fires a navigate action
4. Middleware catches navigate action and uses history api to navigate back to the login page and possibly showing a notification for the user.

I'd rather avoid counting on componentDidUpdate to get the prop which marks it as failed, it can run multiple times, we need to remember to clean it and we're left with data that isn't relevant to the application saved in our stores.

It's much easier to either create a specific middleware that will handle this (for example the middleware could also fire on itself the api action and handle the error itself).

This way we don't have to count on the componentDidUpdate method to run with new props and then decide we want to do something.
In general using componentDidUpdate without comparing the previous value (that it changed values) is very dangerous since it has the potential to cause an infinite loop

Reply:
Perfect, much better. I changed it.